[sparsity] Sparsity parametrization

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60728 [quant][sparsity] Generic convert function
* #59771 [sparsity] Lambda Scheduler
* #59770 [sparsity] Base sparsity level scheduler class
* #58955 [sparsity] WeightNormSparsifier
* #58707 [sparsity][refactor] Import factoring out
* #58704 [sparsity] Sparsifier class
* **#58705 [sparsity] Sparsity parametrization**
* #60850 [sparsity][refactor] Changing linear row/col control
* #60887 [sparsity] Add sparsity tests to run_test.py

The basic demo for this particular implementation can be found here:
https://gist.github.com/z-a-f/1d06ae8d5a509d3c9c1596dcb924afe0

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970959](https://our.internmc.facebook.com/intern/diff/D28970959)

Commit:
[sparsity] Sparse linear layers

[ghstack-poisoned]

Commit:
Update on "[sparsity] Sparse linear layers"




[ghstack-poisoned]

Commit:
Update on "[sparsity] Sparse linear layers"




[ghstack-poisoned]

Commit:
Update on "[sparsity] Sparse linear layers"




[ghstack-poisoned]

Commit:
Update on "[sparsity] Sparse linear layers"




[ghstack-poisoned]

Commit:
Update on "[sparsity] Sparse linear layers"




[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970959).

raghuramank100 commented on torch/ao/nn/sparse/linear.py:
```diff
@@ -0,0 +1,32 @@
+import torch
+from torch.nn import functional as F
+
+class Linear(torch.nn.Linear):
```
We should have the masking done by a module that can operate on the weights. This would be a FakeSparseModule, whos forward method masks the weights.  Can you rewrite this implementation such that:
1. We have a base FakeSparseModule class that contains a mask attribute.
2. The FakeSparseModule is applied to the weight as a parameterization.

raghuramank100 replied:
Sure, in this PR lets do the following:
1. define fakesparse module [similar to the mulby module]
```python 
class FakeSparse(nn.Module):
    def __init__(self, shape):
        super().__init__()
       # Initialize mask to be all ones

        self.mask = torch.ones(shape)

    def forward(self, x):
        assert self.mask.shape == x.shape
        return self.mask * x
```

If you are doing the reparameterization in a later PR, use this module manually, similar to how it is used in nn.qat.linear.
```
 def __init__(self, in_features, out_features, bias=True, device=None, dtype=None) -> None:
        factory_kwargs = {'device': device, 'dtype': dtype}
        super().__init__(in_features, out_features, bias, **factory_kwargs)
        self.fake_sparse = FakeSparse(self.weight.size())

    def forward(self, input):
        return F.linear(input, self.fake_sparse(self.weight), self.bias)
```

raghuramank100 commented on torch/ao/nn/sparse/linear.py:
```diff
@@ -0,0 +1,32 @@
+
+    @classmethod
+    def from_dense(cls, dense):
+        sparse = cls(dense.in_features, dense.out_features, (dense.bias is not None))
```
Can we handle this instead by doing the default init of the mask attr to be all ones? This way from_dense is not needed.

raghuramank100 commented on torch/ao/nn/sparse/quantized/linear.py:
```diff
@@ -82,7 +83,7 @@ class Linear(torch.nn.Module):
     """
     _version = 1
-    _FLOAT_MODULE = torch.nn.Linear
+    _FLOAT_MODULE = (torch.nn.Linear, SparseLinear)
```
I am assuming that nn.Linear is needed for the PTQ flow and SparseLinear for the QAT flow. Is that correct?

raghuramank100 replied:
Lets remove the torch.nn.Linear as a supported option, In both flows, we will need to create a SparseLinearModule. In the Post training sparsity case, one could do:
```python
m = nn.Linear(3,5)
# Sparsifier.prepare would essentially do this:
m_sparse = torch.nn.sparse.linear.from_dense(m)
# m_sparse now has the same weights, but the mask is set to identity
m_sparse.set_mask(mask, pattern) # Function to override the mask and define the sparsity pattern
# Now we can convert only the sparse linear module.
```

raghuramank100 commented on torch/ao/nn/sparse/quantized/linear.py:
```diff
@@ -157,26 +159,24 @@ def set_weight_bias(self, w: torch.Tensor, b: Optional[torch.Tensor],
-        weight = mod.weight
-        if getattr(mod.qconfig, 'mask', False):
-            weight = mod.qconfig.mask * mod.weight
+        weight = mod.weight * mod.mask
```
Let us make mask an attribute of the fakeSparse Module and do this via reparameterization

raghuramank100 commented on torch/ao/nn/sparse/quantized/linear.py:
```diff
@@ -157,26 +159,24 @@ def set_weight_bias(self, w: torch.Tensor, b: Optional[torch.Tensor],
 
     @classmethod
-    def from_float(cls, mod):
+    def from_float(cls, mod, row_block_size=1, col_block_size=4):
```
We should not need the row/block col size here. This block just applies the mask. The mask should enforce the required sparsity pattern

raghuramank100 replied:
I see, from the point of view of keeping the API clean, should we do the following:
1. When we create a sparseLinear module, the fakesparse module should have all the information needed to specify the mask.
2. At convert, we only rely on the fake-sparse module
```python
 def from_float(cls, mod):
  sparsity_pattern = mod.fake_sparse.sparsity_config.sparsity_pattern # This is the shape of the sparse blocks
```

raghuramank100 replied:
Is this being addressed in a later PR?

raghuramank100 reviewed (changes requested):
Please add tests so that functionality is verified.

Commit:
Update on "[sparsity] Sparse linear layers"


Differential Revision: [D28970959](https://our.internmc.facebook.com/intern/diff/D28970959)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970959).

Commit:
Update on "[sparsity] Sparse linear layers"


Demo for the current PR: https://gist.github.com/z-a-f/1d06ae8d5a509d3c9c1596dcb924afe0

Differential Revision: [D28970959](https://our.internmc.facebook.com/intern/diff/D28970959)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970959).

raghuramank100 commented on torch/ao/nn/sparse/quantized/linear.py:
```diff
@@ -157,26 +159,25 @@ def set_weight_bias(self, w: torch.Tensor, b: Optional[torch.Tensor],
 
         We only care about the convert at this stage, no need for observers just yet.
+
+        TODO: Need to figure out how to store the block shapes in the mod
```
We could have an attr in the sparse quantized linear module to store the sparsity_config

raghuramank100 commented on torch/ao/sparsity/parametrization.py:
```diff
@@ -0,0 +1,10 @@
+from torch import nn
```
We are implementing FakeSparsity by parameterization, I see this as the base-class that can be overridden by other implementations of FakeSparse (for example with learnable masks).
Suggest calling this file FakeSparse.py (similar to FakeQuantize.py)

raghuramank100 commented on torch/ao/sparsity/parametrization.py:
```diff
@@ -0,0 +1,10 @@
+
+    def forward(self, x):
+        assert self.other.shape == x.shape
+        return self.other * x
```
Suggest having basic tests here:
1. Create sparseLinearModule, ensure that weights, bias are instantiated correctly [from_dense method works]
2. Ensure that from_float works correctly and the quantized sparse module has the correct weights
3. Ensure that sparseLinearModule can be serialized (both state_dict and jit serialization)
4. Ensure that sparseLinearModule is symbolically traceable.

Commit:
Update on "[sparsity] Sparse linear layers"


Demo for the current PR: https://gist.github.com/z-a-f/1d06ae8d5a509d3c9c1596dcb924afe0

Differential Revision: [D28970959](https://our.internmc.facebook.com/intern/diff/D28970959)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970959).

Commit:
Update on "[sparsity] Sparse linear layers"


Demo for the current PR: https://gist.github.com/z-a-f/1d06ae8d5a509d3c9c1596dcb924afe0

Differential Revision: [D28970959](https://our.internmc.facebook.com/intern/diff/D28970959)

[ghstack-poisoned]

raghuramank100 commented:
Can we also add type annotations to the code for better type checking?

Commit:
Update on "[sparsity] Sparsity parametrization"


The basic demo for this particular implementation can be found here:
https://gist.github.com/z-a-f/1d06ae8d5a509d3c9c1596dcb924afe0

Differential Revision: [D28970959](https://our.internmc.facebook.com/intern/diff/D28970959)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970959).

raghuramank100 commented on torch/ao/nn/sparse/quantized/dynamic/linear.py:
```diff
@@ -55,7 +57,7 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
         op_type = int(state_dict[prefix + 'op_type'])
         assert op_type == 'sparse', \
-               "Cannot load from op_type [{}], expecting [{}]".format(op_type, self._op_type)
+            "Cannot load from op_type [{}], expecting [{}]".format(op_type, self._op_type)
```
Do you need op_type as a field? It is not used anywhere except here.

raghuramank100 commented on torch/ao/nn/sparse/quantized/linear.py:
```diff
@@ -105,7 +104,8 @@ def __init__(self, in_features, out_features, row_block_size, col_block_size, bi
         self.zero_point = 0
 
-    def _get_name(self):
+    @classmethod
```
Can this be just the type(object)? Is this really needed?

raghuramank100 reviewed (changes requested):
Please address remaining comments and add tests

Commit:
Update on "[sparsity] Sparsity parametrization"


The basic demo for this particular implementation can be found here:
https://gist.github.com/z-a-f/1d06ae8d5a509d3c9c1596dcb924afe0

Differential Revision: [D28970959](https://our.internmc.facebook.com/intern/diff/D28970959)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Sparsity parametrization"


The basic demo for this particular implementation can be found here:
https://gist.github.com/z-a-f/1d06ae8d5a509d3c9c1596dcb924afe0

Differential Revision: [D28970959](https://our.internmc.facebook.com/intern/diff/D28970959)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Sparsity parametrization"


The basic demo for this particular implementation can be found here:
https://gist.github.com/z-a-f/1d06ae8d5a509d3c9c1596dcb924afe0

Differential Revision: [D28970959](https://our.internmc.facebook.com/intern/diff/D28970959)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Sparsity parametrization"


The basic demo for this particular implementation can be found here:
https://gist.github.com/z-a-f/1d06ae8d5a509d3c9c1596dcb924afe0

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970959](https://our.internmc.facebook.com/intern/diff/D28970959)

[ghstack-poisoned]

raghuramank100 commented on test/ao/sparsity/test_parametrization.py:
```diff
@@ -0,0 +1,139 @@
+import torch
+from torch.testing._internal.common_utils import TestCase
+
+logging.basicConfig(format='%(asctime)s - %(name)s - %(levelname)s - %(message)s', level=logging.INFO)
```
Is this the way logging is done in pytorch?

z-a-f replied:
yes -- for unittests

raghuramank100 commented on test/ao/sparsity/test_parametrization.py:
```diff
@@ -0,0 +1,139 @@
+        if bias:
+            self.linear = nn.Parameter(torch.zeros_like(self.linear.bias) + 10.0)
+            self.seq[0] = nn.Parameter(torch.zeros_like(self.seq[0].bias) + 20.0)
+            self.seq[0] = nn.Parameter(torch.zeros_like(self.seq[0].bias) + 30.0)
```
Should be [1] instead of [0]

raghuramank100 replied:
Still not changed here?

raghuramank100 commented on test/ao/sparsity/test_parametrization.py:
```diff
@@ -0,0 +1,139 @@
+        self.assertEqual(model_save.seq[1].parametrizations['weight'][0].mask,
+                         model_load.seq[1].parametrizations['weight'][0].mask)
+
+    def test_jit_trace(self):
```
Can you also add a test to check if parameterized models are scriptable?

z-a-f replied:
The parametrizations are not scriptable

raghuramank100 replied:
Please mark a TODO to add tests for symbolic tracing also. We need paramaterization to compose with these APIs.

raghuramank100 commented on test/ao/sparsity/test_parametrization.py:
```diff
@@ -0,0 +1,139 @@
+            self.seq[0] = nn.Parameter(torch.zeros_like(self.seq[0].bias) + 30.0)
+
+        # Make masks stored in the modules for convenience
+        self.linear.register_buffer('mask', torch.eye(16))
```
I dont think you need this in the rest of the code, also since masks are part of the parameterized module, this is confusing. Can you remove this?

z-a-f replied:
This is for convenience of the unittests only. The overall flow does not involve this

raghuramank100 commented on test/ao/sparsity/test_parametrization.py:
```diff
@@ -0,0 +1,139 @@
+        assert not hasattr(model.seq[0], 'parametrizations')
+        assert not hasattr(model.seq[1], 'parametrizations')
+
+        parametrize.register_parametrization(model.linear, 'weight',
```
Instead can we just have 
``` python
   mask = torch.eye(16)
   parametrize.register_parametrization(model.linear, 'weight',
                                             utils.FakeSparsity(mask))
 mask = torch.eye(16)
   parametrize.register_parametrization(model.seq[0], 'weight',
                                             utils.FakeSparsity(mask))
 mask = torch.eye(16)
   parametrize.register_parametrization(model.seq[1], 'weight',
                                             utils.FakeSparsity(mask))
```

raghuramank100 commented on torch/ao/sparsity/utils.py:
```diff
@@ -0,0 +1,20 @@
+
+    Note::
+
+        Once the mask is passed, the variable should not change the id. The
```
NIT: contents instead of contensts.

raghuramank100 commented on torch/ao/sparsity/utils.py:
```diff
@@ -0,0 +1,20 @@
+        not.
+    """
+    def __init__(self, mask):
+        super().__init__()
```
Should we instead init with shape instead of mask?
The mask could be initialized to torch.ones(shape)
``` python
  def __init__(self, shape):
      super().__init__()
      self.register_buffer('mask', torch.ones(shape))
```

If the user wants to override the mask, they can directly set the attribute.

z-a-f replied:
What if the user already has a mask?

raghuramank100 reviewed (changes requested):
Looks good, a few comments

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970959).

Commit:
Update on "[sparsity] Sparsity parametrization"


The basic demo for this particular implementation can be found here:
https://gist.github.com/z-a-f/1d06ae8d5a509d3c9c1596dcb924afe0

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970959](https://our.internmc.facebook.com/intern/diff/D28970959)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970959).

Commit:
Update on "[sparsity] Sparsity parametrization"


The basic demo for this particular implementation can be found here:
https://gist.github.com/z-a-f/1d06ae8d5a509d3c9c1596dcb924afe0

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970959](https://our.internmc.facebook.com/intern/diff/D28970959)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970959).

Commit:
Update on "[sparsity] Sparsity parametrization"


The basic demo for this particular implementation can be found here:
https://gist.github.com/z-a-f/1d06ae8d5a509d3c9c1596dcb924afe0

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970959](https://our.internmc.facebook.com/intern/diff/D28970959)

[ghstack-poisoned]

raghuramank100 commented on test/ao/sparsity/test_parametrization.py:
```diff
@@ -0,0 +1,145 @@
+        self.seq[0].weight = nn.Parameter(torch.zeros_like(self.seq[0].weight) + 2.0)
+        self.seq[1].weight = nn.Parameter(torch.zeros_like(self.seq[1].weight) + 3.0)
+        if bias:
+            self.linear = nn.Parameter(torch.zeros_like(self.linear.bias) + 10.0)
```
self.seq[0] should be self.seq[1]

raghuramank100 reviewed (approved):
Looks good, some previous comments were missed, please take a look before landing.
Also, please add TODOs for scripting and symbolic tracing test coverage.

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970959).

Commit:
Update on "[sparsity] Sparsity parametrization"


The basic demo for this particular implementation can be found here:
https://gist.github.com/z-a-f/1d06ae8d5a509d3c9c1596dcb924afe0

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970959](https://our.internmc.facebook.com/intern/diff/D28970959)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970959).
